kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+                    throw new IllegalStateException("Can't call `next` on a closed iterator.");
+                }
+
+                if (!hasNext()) throw new NoSuchElementException("No next element");
+
+                return actionsIter.get().next();
+            }
+
+            @Override
+            public void close() throws IOException {
+                if (!closed && actionsIter.isPresent()) {
+                    actionsIter.get().close();
+                    actionsIter = Optional.empty();
+                    closed = true;
+                }
+            }
+
+            /**
+             * If the current `actionsIter` has no more elements, this function finds the next
+             * non-empty file in `filesIter` and uses it to set `actionsIter`.
+             */
+            private void tryEnsureNextActionsIterIsReady() {
+                if (actionsIter.isPresent()) {
+                    // This iterator already has a next element, so we can exit early;
+                    if (actionsIter.get().hasNext()) {
+                        return;
+                    }
+
+                    // Clean up resources
+                    try {
+                        actionsIter.get().close();
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+
+                    // Set this to empty since we don't know if there's a next file yet
+                    actionsIter = Optional.empty();
+                }
+
+                // Search for the next non-empty file and use that iter
+                while (filesIter.hasNext()) {
+                    actionsIter = Optional.of(getNextActionsIter());
+
+                    if (actionsIter.get().hasNext()) {
+                        // It is ready, we are done
+                        return;
+                    }
+
+                    // It was an empty file. Clean up resources
+                    try {
+                        actionsIter.get().close();
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+
+                    // Set this to empty since we don't know if there's a next file yet
+                    actionsIter = Optional.empty();
+                }
+            }
+
+            /**
+             * Get the next file from `filesIter` (.json or .checkpoint.parquet), contextualize it
+             * (allow the connector to split it), and then read it + inject the `isFromCheckpoint`
+             * information.
+             *
+             * Requires that `filesIter.hasNext` is true.
+             */
+            private CloseableIterator<Tuple2<FileDataReadResult, Boolean>> getNextActionsIter() {
+                final FileStatus nextFile = filesIter.next();
+                final Path nextFilePath = new Path(nextFile.getPath());
+
+                try {
+                    if (nextFilePath.getName().endsWith(".json")) {
```

Comment:
can we just use the `nextFile.getPath` instead of creating the `Path` object. I think we should try to get rid of the `Path` class completely (or just have some utility classes).